tests/test_utils.py

```diff
@@ -0,0 +1,57 @@
+from __future__ import annotations
+
+from contextlib import nullcontext
+from dataclasses import dataclass
+from typing import Optional, Type, Union
+
+import numpy as np
+import pytest
+
+from instamatic.utils.domains import NumericDomain
+from instamatic.utils.native import AnyNumber, NativeNumber, native
+from tests.utils import InstanceAutoTracker
+
+
+@dataclass
+class NumericDomainTestCase(InstanceAutoTracker):
+    kwargs: dict[str, Union[float, tuple[float, ...]]]
+    raises: Optional[Type[Exception]] = None
+    returns: Optional[float] = None
+
+
+NumericDomainTestCase(kwargs=dict(options=(31, 32, 33)), returns=33)
+NumericDomainTestCase(kwargs=dict(options=(41, 42, 43)), returns=42)
+NumericDomainTestCase(kwargs=dict(options=(51, 52, 53)), returns=51)
+NumericDomainTestCase(kwargs=dict(options=(41, 43, 44)), returns=41)
+NumericDomainTestCase(kwargs=dict(options=(44, 43, 40)), returns=43)
+NumericDomainTestCase(kwargs=dict(lower_lim=31, upper_lim=34), returns=34)
+NumericDomainTestCase(kwargs=dict(lower_lim=41, upper_lim=44), returns=42)
+NumericDomainTestCase(kwargs=dict(lower_lim=51, upper_lim=54), returns=51)
+NumericDomainTestCase(kwargs=dict(lower_lim=51, options=(31, 32)), raises=ValueError)
+NumericDomainTestCase(kwargs=dict(), raises=ValueError)
+
+
+@pytest.mark.parametrize('test_case', NumericDomainTestCase.INSTANCES)
+def test_float_domain(test_case) -> None:
+    """Assert a `FloatDomain` subclass is initialized and works correctly."""
+    c = test_case
+    with pytest.raises(r) if (r := c.raises) else nullcontext():
+        assert NumericDomain(**c.kwargs).nearest(to=42) == c.returns
+
+
+@dataclass
+class NativeTestCase(InstanceAutoTracker):
+    input_value: AnyNumber
+    output_type: Type[NativeNumber]
+
+
+NativeTestCase(input_value=np.float64(1.23), output_type=float)
+NativeTestCase(input_value=np.int64(1), output_type=int)
+NativeTestCase(input_value=float(1.23), output_type=float)
+NativeTestCase(input_value=int(1), output_type=int)
+
+
+@pytest.mark.parametrize('test_case', NativeTestCase.INSTANCES)
```

Comment:
Wow, a bit too much magic here for my tastes, this _is_ a cool way of paramterizing tests.

I wonder if you can make it more magic by implementing `__iter__()` on `InstanceAutoTracker` to turn it into an iterable (https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable) and get rid of the `INSTANCES` attribute.

```suggestion
@pytest.mark.parametrize('test_case', NativeTestCase)
```

Reply:
```python
class IterableType(type):
    def __iter__(self) -> Iterator:
        yield from self.INSTANCES


class InstanceAutoTracker(metaclass=IterableType):
    """Track cls instances: useful for @pytest.mark.parametrize dataclasses"""

    def __init_subclass__(cls, **kwargs) -> None:
        super().__init_subclass__(**kwargs)
        cls.INSTANCES: list[Self] = []

    def __post_init__(self) -> None:
        self.__class__.INSTANCES.append(self)
```

Learning metaclasses in Python opens some door that maybe should remain closed 😎 . I definitely do not think skipping one class variable call is worth added complexity, as cool as it is though.